src/Nerdbank.Streams/NestedStream.cs

```diff
@@ -55,12 +61,23 @@ public NestedStream(Stream underlyingStream, long length)
         public override bool CanWrite => false;
 
         /// <inheritdoc />
-        public override long Length => throw this.ThrowDisposedOr(new NotSupportedException());
+        public override long Length
+        {
+            get
+            {
+                Verify.NotDisposed(this);
+                return this.length;
+            }
+        }
```

Comment:
We don't actually know the length, strictly speaking. We only know the *maximum* length. I can imagine that causing failures in some code that if we claim they can read *x* bytes that they'll be able to before reaching the end of the stream.

Reply:
Good point. My use cases for nested streams are typically archives (tar/cpio/zip/xip/disk images with file systems/...), where the length is (almost) always known but the underlying stream may not be seekable (e.g. if the underlying data is compressed).

I see a couple of options:
1. Always throw `NotSupported`
2. Only report length if the underlying stream is seekable (which allows you to make sure the nested stream does not extend beyond the end of the underlying stream)
3. Always report the length "as is"
4. Delegate the choice to the caller; i.e., have `ReadSlice` and `ReadSliceExact` method (or something similar)

Perhaps 1 ("not my concern") and 4 ("caller knows best") are the better choices?

Reply:
I implemented 2 for now since it doesn't introduce additional complexity in the `Read` methods.